File: Linear Regression/symbolicApproach.py
from sympy import symbols, Matrix
import sympy as sp
import pandas as pd
from tqdm import tqdm

# Function to project vector a onto vector b
def project(a, b):
    return (a.dot(b) / b.dot(b)) * b

# Function to calculate the residual vector
def calculate_residual(label_vector, feature_vectors):
    residual = label_vector
    for col in tqdm(range(feature_vectors.shape[1]), desc="Calculating residual"):
        feature = feature_vectors.col(col) 
        residual -= project(residual, feature)

    return residual

# Function to Compute Row Echelon Form (REF) with symbolic computation
def get_ref_and_pivots(matrix):
    ref_matrix, pivot_columns = matrix.T.rref()
    return ref_matrix.T, pivot_columns

# Function to simplify symbolic matrix
def simplify_matrix(matrix):
    print("Simplifying matrix...")
    return matrix.applyfunc(sp.simplify)

# Load the dataset
file_path = 'datasets/cleaned_air_quality.csv'
df = pd.read_csv(file_path)
df = df.head(20)

print("Replacing missing values with unique symbols...")
# Step 1: Replace missing values with unique symbols
for col in df.columns:
    if df[col].isnull().any():
        symbol = symbols(f'missing_{col}')
        df[col].fillna(symbol, inplace=True)
        
print("DataFrame:")
print(df)

# # Extract label vector and feature matrix
label_vector = Matrix(df['T'].values)
feature_vectors = Matrix(df.drop(columns=['T']).values.tolist())

# Step 2: Compute Row Echelon Form (REF) and get pivot columns
# print("Computing linearly independant features...")
# ref_matrix, pivot_columns = get_ref_and_pivots(feature_vectors)
# independent_features = feature_vectors[:, pivot_columns]

# If not all vectors are linearly independent, exit the program (cannot guarantee minimality)
# Questions: Can we guarantee if all complete features are LI?
# https://github.com/ResidentMario/missingno

# Step 3: Simplify feature vectors
feature_vectors = simplify_matrix(feature_vectors)

print("Simplified feature vectors:")
for row in feature_vectors.tolist():
    print(row)

# Step 4: Calculate the residual vector using the independent features
print("Calculating residual...")
residual_vector = calculate_residual(label_vector, feature_vectors)
residual_vector = simplify_matrix(residual_vector)
print("Simplified Residual vector:", residual_vector.evalf(5), end="\n\n")

# # Step 5: Define vector b
b = label_vector - residual_vector

# # Step 6: Solve the system A * x = b
A = feature_vectors
print("Solving the system A * x = b...")

try:
    # Attempt to solve using least squares 
    # solution = A.solve_least_squares(b, method='CH')    # CH    = Cholesky decomposition
    solution = A.solve_least_squares(b, method='LDL')     # LDL   = LDLsolve
    # solution = A.solve_least_squares(b, method='QR')    # QR    = QRsolve decomposition
    # solution = A.solve_least_squares(b, method='PINV')  # PINV  = pinv_solve decomposition
    # solution = A.solve_least_squares(b)                 # Default

except Exception as e:
    print("Exception", e)
    print("Matrix is non-invertible, using gauss_jordan_solve instead...")
    solution, _ = A.gauss_jordan_solve(b)
    solution = simplify_matrix(solution)

# # Step 7: Pretty print the solution
print("Formatting solution...")
feature_names = df.drop(columns=['T']).columns

# Dictionary of feature : coefficient pairs
solution_dict = {name: coef for name, coef in zip(feature_names, solution)}

# # Step 8: Convert the solution to a DataFrame for better visualization
solution_df = pd.DataFrame(list(solution_dict.items()), columns=['Feature', 'Coefficient'])
print("\nCoefficients DataFrame:")
print(solution_df)


 

File: Linear Regression/symbolicApproach.m


% Load the dataset from the workspace
df = cleanedairquality;

% Limit to the first 500 rows
nRows = min(500, height(df)); % Ensure it does not exceed available rows
df = df(1:nRows, :);

% Display the DataFrame
disp('DataFrame:');
disp(head(df, 10));

disp('Replacing missing values with unique symbols...');
% Convert the table to a cell array to handle symbolic replacements
df_cell = table2cell(df);

% Step 1: Replace missing values with unique symbols
for col = 1:width(df)
    if any(ismissing(df{:, col}))
        symbol = sym(sprintf('missing_%s', df.Properties.VariableNames{col}));
        for row = 1:height(df)
            if ismissing(df{row, col})
                df_cell{row, col} = symbol;
            else
                df_cell{row, col} = sym(df_cell{row, col}); % Convert numerical values to symbolic
            end
        end
    else
        % Convert entire column to symbolic if no missing values
        df_cell(:, col) = num2cell(sym(cell2mat(df_cell(:, col))));
    end
end

% Convert the modified cell array back to a table
df = cell2table(df_cell, 'VariableNames', df.Properties.VariableNames);

disp('Modified DataFrame:');
disp(head(df, 10));

% Extract label vector and feature matrix
label_vector = sym(df{:, strcmp(df.Properties.VariableNames, 'T')});
feature_vectors = sym(df{:, ~strcmp(df.Properties.VariableNames, 'T')});

% Step 3: Calculate the residual vector
disp('Calculating residual...');
residual_vector = calculate_residual(label_vector, feature_vectors);
disp('Residual vector:');
disp(residual_vector);

% Step 4: Define vector b
b = label_vector - residual_vector;

% Step 5: Solve the system A * x = b
A = feature_vectors;
disp('Solving the system A * x = b...');

try
    % Attempt to solve using LU decomposition
    solution = linsolve(A, b);
catch
    disp('Matrix is non-invertible, using pinv instead...');
    solution = pinv(A) * b;
end

% Limit precision of the solution
solution = vpa(solution, 5);

% Step 7: Pretty print the solution
disp('Formatting solution...');
feature_names = df.Properties.VariableNames(~strcmp(df.Properties.VariableNames, 'T'));
solution_dict = containers.Map(feature_names, num2cell(solution));

% Step 8: Convert the solution to a table for better visualization
solution_table = cell2table([keys(solution_dict); values(solution_dict)]');
solution_table.Properties.VariableNames = {'Feature', 'Coefficient'};
disp('Coefficients Table:');
disp(solution_table);

% Define the project function
function proj = project(a, b)
    proj = (dot(a, b) / dot(b, b)) * b;
end

% Define the calculate_residual function
function residual = calculate_residual(label_vector, feature_vectors)
    residual = label_vector;
    for col = 1:size(feature_vectors, 2)
        feature = feature_vectors(:, col);
        residual = residual - project(residual, feature);
    end
end